### ISSUE TITLE
Accidentally raising to a power of a modular number causes infinite loop

### ISSUE THREAD

**Issue by DrPhil** [TikhonJelvis/modular-arithmetic#14]
How to reproduce: evaluate `1 ^ (5 :: Integer / 7)`.

The prelude implementation of `^` uses exponentiation by squaring, and will repeatedly divide by two until the exponent is one. It will even loop on `1 ^ (1 :: Integer / 3)` since `even (1 :: Integer / 3)` is `True`!

**Comment by khalilfazal** [comment]
That gets evaluated as follows:

```haskell
even 1 :: ℤ/3
(1 :: ℤ/3) `rem` 2 == 0
let (_, r) = (1 :: ℤ/3) `quotRem` 2 in r == 0
let (_, r) = (_, 0) in r == 0
0 == 0
True
```

According to this comment, the remainder is supposed to always be zero.
Not sure why.

```haskell
-- | Integer division uses modular inverse @'inv'@, so it is possible
-- to divide only by numbers coprime to @n@ and the remainder is
-- always @0@.
```

**Comment by TikhonJelvis** [comment]
Thanks for finding this. Off-hand, looks like it would be hard to fix since we can't overload `(^)` for our type.

I'm preparing for a talk on Saturday ([shameless plug](https://haskell.love/tikhon-jelvis/)), so I won't have time to look into this until next week. In the meantime, do you have any ideas for what we could do about this problem?

**Comment by khalilfazal** [comment]
Could contact a mathematician specializing in number theory for advice if you know any.  
I wish the maintainers of this module included a link to a paper on Modular Arithmetic on arxiv.org which would justify the definitions chosen.
It seems to be less of an engineering problem and more of a conceptual issue.

I signed up for Haskell Love, I'll wave from the audience. Good luck!

**Comment by TikhonJelvis** [comment]
>  It seems to be less of an engineering problem and more of a conceptual issue.

Yes, I think that's right—it's fundamentally a problem with the `Integral` instance. Seems like the cleanest approach would be to remove the `Integral` instance altogether and use `Fractional` instead, as suggested by @CarlEdman in #8.

Do either of you want to take a crack at it? I'm more than happy to review and merge a PR with that approach or some other design if you have an alternative in mind. Otherwise, I'll make the change myself but, depending on my schedule, probably won't get around to it for a couple of weeks at least.

**Comment by khalilfazal** [comment]
It seems like this is already done in [arithmoi-0.6.0.0](https://github.com/Bodigrim/arithmoi/pull/68)'s `Math.NumberTheory.Moduli.Class::Mod` data type.
See https://github.com/Bodigrim/arithmoi/pull/56.
See the advantages given for that implementation over this one at the bottom of the initial comment.

**Comment by TikhonJelvis** [comment]
Yes, it looks like we can reuse that code with some slight adjustments and it's under a compatible license.

**Comment by TikhonJelvis** [comment]
Okay, this shouldn't be a problem after #15 since we don't have an `Integral` instance any more so using `^` this way won't typecheck.
